File: src/com/nadmm/airports/afd/NearbyAirportsFragment.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011-2012 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

package com.nadmm.airports.afd;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.location.Location;
import android.os.AsyncTask;
import android.os.Bundle;

import com.nadmm.airports.DatabaseManager;
import com.nadmm.airports.DatabaseManager.LocationColumns;

public class NearbyAirportsFragment extends AirportListFragmentBase {

    @Override
    public void onActivityCreated( Bundle savedInstanceState ) {
        Bundle args = getArguments();
        Location location = (Location) args.get( LocationColumns.LOCATION );
        if ( location != null ) {
            onLocationChanged( location );
        }
        getListView().setCacheColorHint( 0xffffffff );

        super.onActivityCreated( savedInstanceState );
    }

    private final class NearbyAirportsTask extends AsyncTask<Location, Void, Cursor> {

        @Override
        protected Cursor doInBackground( Location... params ) {
            if ( getActivity() == null ) {
                cancel( false );
                return null;
            }

            Location location = params[ 0 ];
            Bundle args = getArguments();
            int radius = args.getInt( LocationColumns.RADIUS );
            SQLiteDatabase db = getDatabase( DatabaseManager.DB_FADDS );

            return new NearbyAirportsCursor( db, location, radius );
        }

        @Override
        protected void onPostExecute( Cursor c ) {
            setCursor( c );
        }

    }

    @Override
    public void onCreate( Bundle savedInstanceState ) {
        setEmptyText( "No airports found nearby." );
        super.onCreate( savedInstanceState );
    }

    @Override
    public void onLocationChanged( Location location ) {
        new NearbyAirportsTask().execute( location );
    }

}

File: src/main/java/com/nadmm/airports/afd/NearbyAirportsFragment.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nadmm.airports.afd

import android.content.Intent
import android.database.Cursor
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import com.nadmm.airports.LocationListFragment2
import com.nadmm.airports.data.DatabaseManager
import com.nadmm.airports.data.DatabaseManager.Airports
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class NearbyAirportsFragment : LocationListFragment2() {

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        setEmptyText("No airports found nearby.")

        if (!isLocationUpdateEnabled) {
            setActionBarTitle("Nearby Airports")
            setActionBarSubtitle("Within $nearbyRadius NM radius")
        }
    }

    override fun startLocationTask() {
        viewLifecycleOwner.lifecycleScope.launch {
            val result = withContext(Dispatchers.IO) {
                doQuery()
            }
            setCursor(result[0])
        }
    }

    override fun newListAdapter(cursor: Cursor?): AirportsRecyclerViewAdapter? {
        return if (cursor != null) {
            AirportsRecyclerViewAdapter(cursor, ::onRecyclerItemClick)
        } else {
            null
        }
    }

    private fun onRecyclerItemClick(model: AirportListDataModel) {
        Intent(activity, AirportActivity::class.java).apply {
            putExtras(model.makeBundle())
            startActivity(this)
        }
    }

    private fun doQuery(): Array<Cursor> {
        val db = getDatabase(DatabaseManager.DB_FADDS)

        val extraSelection = arguments?.getString(Airports.SITE_NUMBER)?.let {
            "AND ${Airports.SITE_NUMBER} <> '$it'"
        }

        val c = NearbyAirportsCursor(db, lastLocation!!, nearbyRadius, extraSelection)
        return arrayOf(c)
    }

}
